server/grpc_service.go

```diff
@@ -0,0 +1,505 @@
+	}
+
+	log.Infof("put store ok - %v", store)
+
+	return &pdpb2.PutStoreResponse{
+		Header: s.header(),
+	}, nil
+}
+
+// StoreHeartbeat implements gRPC PDServer.
+func (s *Server) StoreHeartbeat(ctx context.Context, request *pdpb2.StoreHeartbeatRequest) (*pdpb2.StoreHeartbeatResponse, error) {
+	if !s.IsLeader() {
+		// TODO: work as proxy.
+		return nil, grpc.Errorf(codes.Unavailable, "not leader")
+	}
+	if request.GetHeader().GetClusterId() != s.clusterID {
+		return nil, grpc.Errorf(codes.FailedPrecondition, "mismatch cluster id, need %d but got %d", s.clusterID, request.GetHeader().GetClusterId())
+	}
+	if request.GetStats() == nil {
+		return nil, errors.Errorf("invalid store heartbeat command, but %v", request)
+	}
+	cluster := s.GetRaftCluster()
+	if cluster == nil {
+		return &pdpb2.StoreHeartbeatResponse{Header: s.notBootstrappedHeader()}, nil
+	}
+
+	if pberr := checkStore2(cluster, request.GetStats().GetStoreId()); pberr != nil {
+		return &pdpb2.StoreHeartbeatResponse{
+			Header: s.errorHeader(pberr),
+		}, nil
+	}
+
+	stats := &pdpb.StoreStats{
+		StoreId:            request.Stats.StoreId,
+		Capacity:           request.Stats.Capacity,
+		Available:          request.Stats.Available,
+		RegionCount:        request.Stats.RegionCount,
+		SendingSnapCount:   request.Stats.SendingSnapCount,
+		ReceivingSnapCount: request.Stats.ReceivingSnapCount,
+		StartTime:          request.Stats.StartTime,
+		ApplyingSnapCount:  request.Stats.ApplyingSnapCount,
+		IsBusy:             request.Stats.IsBusy,
+	}
+
+	err := cluster.cachedCluster.handleStoreHeartbeat(stats)
+	if err != nil {
+		return nil, grpc.Errorf(codes.Unknown, err.Error())
+	}
+
+	return &pdpb2.StoreHeartbeatResponse{
+		Header: s.header(),
+	}, nil
+}
+
+func toPDPB2PeerStats(stats []*pdpb2.PeerStats) []*pdpb.PeerStats {
+	ret := make([]*pdpb.PeerStats, 0, len(stats))
+	for _, s := range stats {
+		n := &pdpb.PeerStats{
+			Peer: s.Peer,
+		}
+		if s.DownSeconds != 0 {
+			n.DownSeconds = proto.Uint64(s.DownSeconds)
+		}
+		ret = append(ret, n)
+	}
+	return ret
+}
+
+// RegionHeartbeat implements gRPC PDServer.
+func (s *Server) RegionHeartbeat(ctx context.Context, request *pdpb2.RegionHeartbeatRequest) (*pdpb2.RegionHeartbeatResponse, error) {
+	if !s.IsLeader() {
+		// TODO: work as proxy.
+		return nil, grpc.Errorf(codes.Unavailable, "not leader")
+	}
+	if request.GetHeader().GetClusterId() != s.clusterID {
+		return nil, grpc.Errorf(codes.FailedPrecondition, "mismatch cluster id, need %d but got %d", s.clusterID, request.GetHeader().GetClusterId())
+	}
+	cluster := s.GetRaftCluster()
+	if cluster == nil {
+		return &pdpb2.RegionHeartbeatResponse{Header: s.notBootstrappedHeader()}, nil
+	}
+
+	resp := &pdpb2.RegionHeartbeatResponse{}
+	region := newRegionInfo(request.GetRegion(), request.GetLeader())
+	region.DownPeers = toPDPB2PeerStats(request.GetDownPeers())
+	region.PendingPeers = request.GetPendingPeers()
+	if region.GetId() == 0 {
+		pberr := &pdpb2.Error{
```

Comment:
It's copied from @overvenus 's branch: https://github.com/pingcap/pd/blob/ov/grpc-pdpb/server/api/v2rpc/server.go
It may not work well with tikv-server if we change it. /cc @overvenus